DOCS-539: Refactor Color follower (#1662)
viamrobotics/docs: docs/tutorials/services/color-detection-scuttle.md

viamresources: [ "base", "vision", "camera" ]
level: "Beginner"
date: "18 August 2022"
updated: "11 August 2023"

> cost: 540

Author: ```suggestion
cost: 540
no_list: true
```

</div>

You can see the [full code](#full-code) at the bottom of the page.

## Requirements

You will need the following hardware to complete this tutorial:

- A wheeled rover, configured with a [base component](/components/base/) on the [Viam app](https://app.viam.com/).
  This tutorial uses a [SCUTTLE rover](https://www.scuttlerobot.org/shop/) as an example but you can complete this tutorial using a different rover.
  - Regardless of the type of base you are using, [Setting up a SCUTTLE with Viam](/tutorials/configure/scuttlebot/) is a good place to start if you haven't already configured your base.
- An attached and configured [webcam camera](/components/camera/webcam/).

## Set up the hardware

Connect the camera to the rover's board.
Turn on the power to the rover.

## Configure color detection

This tutorial uses the color `#a13b4c` or `rgb(161,59,76)` (a reddish color).

{{< tabs >}}
{{% tab name="Builder" %}}

Navigate to the [robot page on the Viam app](https://app.viam.com/robots).
Click on the robot you wish to add the vision service to.
Select the **Config** tab, and click on **Services**.

Scroll to the **Create Service** section.

> To create a [color detector vision service](/services/vision/detector):

Author: ```suggestion
To create a [color detector vision service](/services/vision/detection/):
```

3. Navigate to the **Control** tab, click on your transform camera and toggle it on.
   The transform camera will now show detections with bounding boxes around the detected colors.

{{< alert title="Tip" color="tip" >}}
If the color is not reliably detected, increase the `hue_tolerance_pct`.

Note that the detector does not detect black, perfect greys (greys where the red, green, and blue color component values are equal), or white.
{{< /alert >}}

## Program your rover

### Set up your code environment

We are going to use Virtualenv to set up a virtual environment for this project, in order to isolate the dependencies of this project from other projects.
Run the following commands in your command-line to install virtualenv, set up an environment `venv` and activate it:

```sh {class="command-line" data-prompt="$"}
python3 -m pip install --user virtualenv
python3 -m venv env
source env/bin/activate
```

Now, install the Python Viam SDK:

```sh {class="command-line" data-prompt="$"}
pip3 install viam-sdk python-vlc
```

### Connect

Next, go to the **Code sample** tab on your robot page and select **Python**, then click **Copy**.

{{% snippet "show-secret.md" %}}

> This code snippet imports all the necessary packages and sets up a connection with the Viam app in the cloud.

Reviewer: ```suggestion
This code snippet imports all the necessary packages and sets up a connection with the Viam app.
```

Next, create a file named <file>main.py</file> and paste the boilerplate code from the **Code sample** tab of the Viam app into your file.
Then, save your file.

> Run the code to verify that the Viam SDK is properly installed and that the `viam-server` instance on your robot is live.

Reviewer: Ah, looks like a running instance of `viam-server` is also a requirement: it looks like, for _Configure color detection_, we assume an existing robot, which assumes an existing `viam-server` installation.

I think we'd be fine here just adding a link here, maybe:

```suggestion
Run the code to verify that the Viam SDK is properly installed and that the `viam-server` instance on your robot is live.
If you haven't yet installed `viam-server`, follow the [installation guide](/installation/#install-viam-server) to install `viam-server` on your robot before proceeding with this tutorial.
```

> You can run your code by typing the following into your terminal:

Reviewer: ```suggestion
You can run your code by typing the following into your terminal from the same directory as your `main.py` file:
```

```sh {class="command-line" data-prompt="$"}
python3 main.py
```

The program prints a list of robot resources.

On top of the packages that the code sample snippet imports, add the `random` and the `vlc` package to the imports.
The top of your code should now look like this:

```python {class="line-numbers linkable-line-numbers"}
import asyncio
import random
import vlc

from viam.robot.client import RobotClient
from viam.rpc.dial import Credentials, DialOptions
from viam.components.board import Board
from viam.components.camera import Camera
from viam.components.servo import Servo
from viam.services.vision import VisionClient

async def connect():
    creds = Credentials(
        type="robot-location-secret",
        payload="LOCATION SECRET FROM THE VIAM APP")
    opts = RobotClient.Options(
        refresh_interval=0,
        dial_options=DialOptions(credentials=creds)
    )
    return await RobotClient.at_address("ADDRESS FROM THE VIAM APP", opts)

async def main():
    # Other code

if __name__ == "__main__":
  print("Starting up... ")
  asyncio.run(main())
  print("Done.")
```

You will update the `main()` function later.

> ### Detect where the colored object is

Reviewer: nit: to avoid the trailing article

```suggestion
### Detect the location of a colored object
```

Each detection comes with information about where in the camera's picture it is detected.

The following `leftOrRight` function checks where in the picture the largest detection is and returns a responding integer: `0` for left, `1` for center, and `2` for right.

Add the `leftOrRight` function below your `connect` function:

```python {class="line-numbers linkable-line-numbers"}
# Get largest detection box and see if it's center is in the left, center, or right third
def leftOrRight(detections, midpoint):
   largest_area = 0
   largest = Detection()
   if not detections:
      print("nothing detected :(")
      return -1
   for d in detections:
      a = (d.x_max - d.x_min) * (d.y_max-d.y_min)
      if a > largest_area:
         a = largest_area
         largest = d
   centerX = largest.x_min + largest.x_max/2
   if centerX < midpoint-midpoint/6:
      return 0 # on the left
   if centerX > midpoint+midpoint/6:
      return 2 # on the right
   else:
      return 1  # basically centered
```

### Add the main function

The `main` function:

- defines variables for how the robot should move,
- connects to the robot,
- initializes the base, the camera, and the detector, and

> - repeatedly calls the `leftOrRight` function and turns the rover's base into the respective direction.

Reviewer: ```suggestion
- repeatedly calls the `leftOrRight` function and turns the rover's base in the respective direction.
```

   numCycles = 200      # run the loop X times
   vel = 500            # go this fast when moving motor
​
   # Connect to robot client and set up components
   robot = await connect()
   base = Base.from_robot(robot, "base")
   camera = Camera.from_robot(robot, "<camera-name>")

   # Grab the vision service for the detector
   my_detector = VisionClient.from_robot(robot, "my_color_detector")

   # Main loop. Detect the ball, determine if it's on the left or right, and head that way.
   # Repeat this for numCycles
   for i in range(numCycles):
      detections = await my_detector.get_detections_from_camera(camera)
​
      answer = leftOrRight(detections, frame.size[0]/2)
      if answer == 0:
         print("left")
         await base.spin(spinNum, vel)  # CCW is positive
         await base.move_straight(straightNum, vel)
      if answer == 1:
         print("center")
         await base.move_straight(straightNum, vel)
      if answer == 2:
         print("right")
         await base.spin(-spinNum, vel)
      # If nothing is detected, nothing moves

   await robot.close()​
```

For `<camera-name>`, insert the name of your configured physical camera.

## Run the code

> Now, run the code:

Reviewer: ```suggestion
Now, run the code again, from the same directory as your `main.py` file:
```

```sh {class="command-line" data-prompt="$"}
python3 main.py
```

> If everything works, your rover should detect and navigate towards red objects.

Reviewer: opt: confidence & clarification

```suggestion
Your rover should detect and navigate towards any red objects that come into view of its camera.
```

Use something like a red sports ball or book cover as a target to follow to test your rover:

<div class="aligncenter">
{{<video webm_src="/tutorials/videos/scuttledemos_colordetection.webm" mp4_src="/tutorials/videos/scuttledemos_colordetection.mp4" poster="/tutorials/scuttlebot/scuttledemos_colordetection.jpg" alt="Detecting color with a Scuttle Robot">}}
</div>

## Next Steps

Congratulations! If you're ready for more, try making your rover detect other colors.
You could also write some code with a Viam SDK to [make your rover move in a square](/tutorials/get-started/try-viam-sdk/).

> {{< snippet "social.md" >}}

Reviewer: nit: technically this also appears at the bottom, below full code. I'm OK with it though.

   For `<camera-name>`, insert the name of your configured physical camera.

2. Click **Save config**.
3. Navigate to the **Control** tab, click on your transform camera and toggle it on.
   The transform camera will now show detections with bounding boxes around the detected colors.

{{< alert title="Tip" color="tip" >}}
If the color is not reliably detected, try increasing the `hue_tolerance_pct` or adjusting the lighting of the area to make the color being detected more visible.

Note that the detector does not detect black, perfect greys (greys where the red, green, and blue color component values are equal), or white.
{{< /alert >}}

## Program your rover

### Set up your code environment

We are going to use Virtualenv to set up a virtual environment for this project, in order to isolate the dependencies of this project from other projects.
Run the following commands in your command-line to install `virtualenv`, set up an environment `venv` and activate it:

```sh {class="command-line" data-prompt="$"}
python3 -m pip install --user virtualenv
python3 -m venv env
source env/bin/activate
```

Then, install the Viam Python SDK:

```sh {class="command-line" data-prompt="$"}

> pip3 install viam-sdk python-vlc

Author: ```suggestion
pip3 install viam-sdk
```
